test/upgrade/main.go

```diff
@@ -336,8 +336,47 @@ func runConfigWalker(ctx context.Context, validConfigs []*configTest) {
 		time.Sleep(1 * time.Minute)
 	}
 	cancel()
-	// TODO: Replace sleep with wait for the existing healthy Game Servers finish naturally by reaching their shutdown phase.
-	time.Sleep(30 * time.Second)
+
+	waitForGameServersToShutdown(ctx, agonesClient)
+}
+
+// waitForGameServersToShutdown waits for all existing healthy Game Servers to reach their shutdown phase
+func waitForGameServersToShutdown(ctx context.Context, agonesClient *versioned.Clientset) {
+	log.Println("Waiting for existing Game Servers to shutdown...")
+
+	err := wait.PollUntilContextTimeout(ctx, 5*time.Second, 2*time.Minute, true, func(ctx context.Context) (done bool, err error) {
+		// Get all Game Servers with the test label
+		gameServers, err := agonesClient.AgonesV1().GameServers(PodNamespace).List(ctx, metav1.ListOptions{
+			LabelSelector: "app=sdk-client-test",
+		})
+		if err != nil {
+			log.Printf("Error listing Game Servers: %v", err)
+			return false, nil
+		}
+
+		// Count Game Servers that are still in healthy states (not shutdown/terminated)
+		healthyCount := 0
+		for _, gs := range gameServers.Items {
+			state := gs.Status.State
+			// Consider Game Servers healthy if they're in Ready, Allocated, or Reserved states
+			if state == agonesv1.GameServerStateReady ||
+				state == agonesv1.GameServerStateAllocated ||
+				state == agonesv1.GameServerStateReserved {
+				healthyCount++
+			}
+		}
+
+		log.Printf("Waiting for %d healthy Game Servers to shutdown", healthyCount)
+
+		// Return true when no healthy Game Servers remain
+		return healthyCount == 0, nil
```

Comment:
```suggestion
  return len(gameServers) == 0, nil
```

Is what we're saying?

Out of curiosity - why do we wait for GameServers to be totally deleted?

Reply:
It's a guess, but the issue could be that:

We delete the GS using kubectl, they are in shutdown etc. Them we delete the helm release without waiting that it's done, which start but never finish and fails on delete namespace as there is still resources on it ? 

The issue being that we dont wait between the delete gs and the unisntall of the helm release, we might set a finalizers and it never be deleted  as the helm uninstall doesn't wait that the resources are cleaned up, this might be the explanation about why we want to wait that they are either non ready or deleted

Reply:
@igooch correct me if I'm wrong, but does that defeat the purpose of the test - which is to test in place upgrades.

Reply:
From what I understood it's only after the test that runs the in place upgrades ([runConfigWalker](https://github.com/googleforgames/agones/blob/main/test/upgrade/main.go#L300)) which is called just before the cleanup: [here](https://github.com/googleforgames/agones/blob/main/test/upgrade/main.go#L105), at the really end, once the test is done 

But I might have missed something or miss-understood some part 😄